File: backend/schemas/strategy_module.py
"""Pydantic request/response schemas for the strategy module CRUD API.

Strict mode (``extra='forbid'``) is on every schema — unknown fields are
rejected with 422 instead of silently dropped (Section 14.5 of the plan).

Response timestamps are rendered in IST ISO 8601 via
:func:`backend.strategy.time_utils.format_ist`.
"""

from __future__ import annotations

from datetime import datetime, time
from typing import List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from backend.strategy.time_utils import format_ist

# ----------------------------------------------------------------------------
# Leg / sub-objects
# ----------------------------------------------------------------------------


class TrailConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")

    x: float = Field(0, ge=0, description="Favorable move (pts) before trail arms")
    y: float = Field(0, ge=0, description="Trail step (pts) once armed")


class Leg(BaseModel):
    """One leg in a strategy.

    Validation rules per the plan (Section 4.1.1) plus the signal-mode
    extension (docs/plan/strategy-signal-mode.md):

    Batch mode legs (parent ``strategy_kind="batch"``):
      * ``segment="options"`` requires ``option_type`` and ``strike_mode``.
      * ``strike_mode="atm"`` requires ``atm_offset``.
      * ``strike_mode="strike"`` requires ``strike_value``.

    Signal mode legs (parent ``strategy_kind="signal"``):
      * Each leg carries its own ``symbol`` + ``exchange``.
      * ``side`` declares which signals are accepted.
      * ``qty`` is the absolute order quantity (lotsize already multiplied
        for FUT; raw shares for cash).
      * ``segment`` must be ``"cash"`` or ``"futures"`` (no options in v1).
      * Options-only fields (``option_type``, ``strike_*``) must be unset.

    ``expiry`` is optional - required only for futures (signal mode) and
    F&O batch legs. Cash legs leave it None.

    Per-kind validation is enforced at the parent StrategyCreate /
    StrategyUpdate level so a Leg validates standalone in any context.
    """

    model_config = ConfigDict(extra="forbid")

    id: int = Field(..., ge=1, description="1-indexed leg number, unique within strategy")
    segment: Literal["options", "futures", "cash"]
    # Canonical ranks: current_week / next_week / current_month / next_month.
    # Legacy aliases (kept so existing DB rows keep parsing): weekly = current_week,
    # monthly = current_month, current = current_month, next = next_month.
    # Strategy-level validator below blocks current_week/next_week/weekly on MCX.
    expiry: Optional[Literal[
        "current_week", "next_week", "current_month", "next_month",
        "weekly", "monthly", "current", "next",
    ]] = None
    # Batch-mode lot count (multiplied by symtoken.lotsize at runtime). For
    # signal-mode legs ``qty`` carries the absolute order quantity instead.
    lots: int = Field(1, gt=0, le=10000)
    position: Literal["B", "S"] = "B"

    option_type: Optional[Literal["CE", "PE"]] = None
    strike_mode: Optional[Literal["atm", "strike"]] = None
    atm_offset: Optional[str] = Field(
        None,
        pattern=r"^(ATM|ATM[+-]\d+|ITM\d+|OTM\d+)$",
        description="ATM, ATM+1, ATM-1, ITM2, OTM3, etc.",
    )
    strike_value: Optional[float] = Field(None, gt=0)

    # --- Signal-mode fields (None for batch-mode legs) ---
    symbol: Optional[str] = Field(None, min_length=1, max_length=50)
    exchange: Optional[str] = Field(None, min_length=1, max_length=20)
    side: Optional[Literal["long", "short", "both"]] = None
    qty: Optional[int] = Field(None, gt=0, le=1_000_000)

    target_pts: Optional[float] = Field(None, ge=0)
    sl_pts: Optional[float] = Field(None, ge=0)
    trail: TrailConfig = Field(default_factory=TrailConfig)

    momentum: Optional[dict] = Field(default=None, description="v1 stub — not evaluated")

    @model_validator(mode="after")
    def _validate_segment_fields(self) -> "Leg":
        if self.segment == "options":
            # Options legs are batch-mode only - signal-mode rejects this at
            # the strategy validator.
            if self.option_type is None:
                raise ValueError("option_type required when segment='options'")
            if self.strike_mode is None:
                raise ValueError("strike_mode required when segment='options'")
            if self.strike_mode == "atm" and not self.atm_offset:
                raise ValueError("atm_offset required when strike_mode='atm'")
            if self.strike_mode == "strike" and self.strike_value is None:
                raise ValueError("strike_value required when strike_mode='strike'")
        else:
            # Non-options legs must not carry option-only fields.
            if self.option_type is not None:
                raise ValueError("option_type only valid when segment='options'")
            if self.strike_mode is not None:
                raise ValueError("strike_mode only valid when segment='options'")
        return self


class LockProfitConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")

    mode: Literal["lock", "lock_and_trail"]
    if_profit_reaches: float = Field(..., gt=0)
    lock_profit: float = Field(..., ge=0)
    trail_step: Optional[float] = Field(None, gt=0)

    @model_validator(mode="after")
    def _trail_step_required_for_lock_and_trail(self) -> "LockProfitConfig":
        if self.mode == "lock_and_trail" and self.trail_step is None:
            raise ValueError("trail_step required when mode='lock_and_trail'")
        if self.lock_profit > self.if_profit_reaches:
            raise ValueError("lock_profit must be <= if_profit_reaches")
        return self


class SchedulerConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")

    enabled: bool = False
    days: List[Literal["MON", "TUE", "WED", "THU", "FRI", "SAT", "SUN"]] = Field(
        default_factory=lambda: ["MON", "TUE", "WED", "THU", "FRI"]
    )
    start_time: str = Field("09:15", pattern=r"^([01]\d|2[0-3]):[0-5]\d$")
    auto_stop_time: Optional[str] = Field(None, pattern=r"^([01]\d|2[0-3]):[0-5]\d$")
    default_mode: Literal["live", "sandbox"] = "sandbox"


class WebhookIpAllowlistEntry(BaseModel):
    model_config = ConfigDict(extra="forbid")

    cidr: str = Field(..., min_length=1, max_length=43)
    label: Optional[str] = Field(None, max_length=100)


# ----------------------------------------------------------------------------
# Create / Update / Out
# ----------------------------------------------------------------------------


_STRATEGY_TYPE = Literal["intraday", "positional"]
_UNIVERSE_TAB = Literal["weekly_monthly", "monthly_only", "stocks_fno", "mcx"]
_PRODUCT = Literal["NRML", "MIS", "CNC"]
_PRICETYPE = Literal["MARKET", "LIMIT"]
_STRATEGY_KIND = Literal["batch", "signal"]
_DIRECTION = Literal["long_only", "short_only", "both"]

# MCX commodity contracts are monthly-only (verified from symtoken: every
# MCX FUT/OPT row is a distinct monthly expiry). Weekly ranks are nonsense
# there, so reject them at the schema layer instead of letting the engine
# resolve to an unexpected contract.
_WEEKLY_RANKS_FORBIDDEN_ON_MCX = frozenset(
    {"current_week", "next_week", "weekly"}
)


def _validate_product_vs_segments(product: str, legs: List[Leg]) -> None:
    """Enforce exchange-side product rules against the leg composition.

    Rules (per Indian-exchange conventions):
      * ``CNC``  — delivery, cash equity only. Reject if any leg is FUT/OPT.
      * ``NRML`` — normal margin, derivatives only. Reject if any leg is cash.
      * ``MIS``  — intraday, valid for every segment.
    """
    has_cash = any(leg.segment == "cash" for leg in legs)
    has_deriv = any(leg.segment in ("futures", "options") for leg in legs)
    if product == "CNC" and has_deriv:
        raise ValueError(
            "product='CNC' is delivery-only — not valid for futures or "
            "options legs. Use 'NRML' for derivatives or 'MIS' for a mixed "
            "intraday basket."
        )
    if product == "NRML" and has_cash:
        raise ValueError(
            "product='NRML' is a derivatives product — not valid for cash "
            "equity legs. Use 'CNC' for delivery or 'MIS' for intraday."
        )


def _validate_tab_expiries(universe_tab: str, legs: List[Leg]) -> None:
    """Cross-field validation: per-tab expiry restrictions.

    Today's rule: MCX rejects weekly ranks on any leg (FUT or OPT). NSE/BSE
    index tabs accept the full four ranks. Stocks F&O accepts monthly ranks
    only (no weeklies for stock options/futures even on NFO/BFO).
    """
    if universe_tab == "mcx":
        for leg in legs:
            if leg.expiry in _WEEKLY_RANKS_FORBIDDEN_ON_MCX:
                raise ValueError(
                    f"Leg {leg.id}: MCX contracts are monthly-only — "
                    f"'{leg.expiry}' is not a valid expiry rank for MCX. "
                    f"Use 'current_month' or 'next_month'."
                )
    elif universe_tab == "stocks_fno":
        for leg in legs:
            if leg.expiry in ("current_week", "next_week", "weekly"):
                raise ValueError(
                    f"Leg {leg.id}: stock F&O contracts are monthly-only — "
                    f"weekly ranks aren't valid on '{universe_tab}'. "
                    f"Use 'current_month' or 'next_month'."
                )


def _validate_legs_for_kind(kind: str, legs: List[Leg]) -> None:
    """Per-kind leg validation. Called from StrategyCreate and StrategyUpdate.

    Batch mode: legs must NOT carry signal-mode fields (``symbol``,
    ``exchange``, ``side``, ``qty``). Options legs are allowed.

    Signal mode: each leg MUST carry ``symbol``, ``exchange``, ``side``,
    ``qty``. ``segment`` may be ``cash``, ``futures``, or ``options``;
    the leg's ``symbol`` carries the base underlying (e.g. NIFTY, RELIANCE,
    CRUDEOIL) and the engine resolves the actual FUT/option contract at
    signal time from the leg's expiry rank + option fields.
    """
    if kind == "signal":
        for leg in legs:
            if not leg.symbol:
                raise ValueError(f"Leg {leg.id}: signal-mode legs require 'symbol'")
            if not leg.exchange:
                raise ValueError(f"Leg {leg.id}: signal-mode legs require 'exchange'")
            if leg.side is None:
                raise ValueError(
                    f"Leg {leg.id}: signal-mode legs require 'side' "
                    f"(long / short / both)"
                )
            if leg.qty is None:
                raise ValueError(f"Leg {leg.id}: signal-mode legs require 'qty'")
            if leg.segment == "futures" and leg.expiry not in (
                "current", "next", "current_month", "next_month",
            ):
                raise ValueError(
                    f"Leg {leg.id}: futures legs in signal mode need expiry "
                    f"'current_month' or 'next_month' (legacy 'current'/'next' also accepted)"
                )
            if leg.segment == "options" and leg.expiry is None:
                raise ValueError(
                    f"Leg {leg.id}: options legs in signal mode need expiry "
                    f"(current_week / next_week / current_month / next_month)"
                )
    else:  # batch
        for leg in legs:
            for field in ("symbol", "exchange", "side", "qty"):
                if getattr(leg, field) is not None:
                    raise ValueError(
                        f"Leg {leg.id}: '{field}' is only valid for "
                        f"signal-mode strategies"
                    )
            if leg.expiry is None and leg.segment != "cash":
                raise ValueError(
                    f"Leg {leg.id}: batch-mode F&O legs require 'expiry'"
                )


class StrategyCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(..., min_length=1, max_length=200)
    strategy_kind: _STRATEGY_KIND = "batch"
    direction: _DIRECTION = "both"

    universe_tab: _UNIVERSE_TAB
    underlying: str = Field(..., min_length=1, max_length=50)
    underlying_exchange: str = Field(..., min_length=1, max_length=20)

    strategy_type: _STRATEGY_TYPE
    entry_time: Optional[time] = None
    exit_time: Optional[time] = None

    product: _PRODUCT = "NRML"
    pricetype: _PRICETYPE = "MARKET"

    legs: List[Leg] = Field(..., min_length=1, max_length=10)

    overall_sl_mtm: Optional[float] = Field(None, ge=0)
    overall_target_mtm: Optional[float] = Field(None, ge=0)
    lock_profit: Optional[LockProfitConfig] = None
    trail_sl_to_entry: bool = False

    scheduler: Optional[SchedulerConfig] = None

    webhook_ip_allowlist: Optional[List[WebhookIpAllowlistEntry]] = None
    daily_loss_limit_inr: Optional[float] = Field(None, gt=0)

    @model_validator(mode="after")
    def _validate_intraday_times(self) -> "StrategyCreate":
        if self.strategy_type == "intraday":
            if self.entry_time is None or self.exit_time is None:
                raise ValueError("entry_time and exit_time required for intraday strategies")
            if self.entry_time >= self.exit_time:
                raise ValueError("entry_time must be before exit_time")
        return self

    @model_validator(mode="after")
    def _validate_legs_per_kind(self) -> "StrategyCreate":
        _validate_legs_for_kind(self.strategy_kind, self.legs)
        _validate_tab_expiries(self.universe_tab, self.legs)
        _validate_product_vs_segments(self.product, self.legs)
        return self

    @field_validator("legs")
    @classmethod
    def _unique_leg_ids(cls, legs: List[Leg]) -> List[Leg]:
        ids = [leg.id for leg in legs]
        if len(set(ids)) != len(ids):
            raise ValueError("leg ids must be unique within a strategy")
        return legs


class StrategyUpdate(BaseModel):
    """Partial update — every field optional. PATCH-style.

    PATCH is rejected (409) when ``status != 'stopped'`` at the router layer.
    Mass-assignment of internal fields like ``webhook_token_hash``, ``user_id``,
    ``status``, ``current_run_id`` is impossible because they're not in this
    schema (Pydantic strict mode rejects them).

    Note: ``strategy_kind`` cannot be changed after creation - flipping a
    live strategy from batch to signal (or back) would invalidate the legs
    array and the webhook payload contract. The router enforces this.
    """

    model_config = ConfigDict(extra="forbid")

    name: Optional[str] = Field(None, min_length=1, max_length=200)
    direction: Optional[_DIRECTION] = None
    universe_tab: Optional[_UNIVERSE_TAB] = None
    underlying: Optional[str] = Field(None, min_length=1, max_length=50)
    underlying_exchange: Optional[str] = Field(None, min_length=1, max_length=20)

    strategy_type: Optional[_STRATEGY_TYPE] = None
    entry_time: Optional[time] = None
    exit_time: Optional[time] = None

    product: Optional[_PRODUCT] = None
    pricetype: Optional[_PRICETYPE] = None

    legs: Optional[List[Leg]] = Field(None, min_length=1, max_length=10)

    overall_sl_mtm: Optional[float] = Field(None, ge=0)
    overall_target_mtm: Optional[float] = Field(None, ge=0)
    lock_profit: Optional[LockProfitConfig] = None
    trail_sl_to_entry: Optional[bool] = None

    scheduler: Optional[SchedulerConfig] = None

    webhook_ip_allowlist: Optional[List[WebhookIpAllowlistEntry]] = None
    daily_loss_limit_inr: Optional[float] = Field(None, gt=0)

    @field_validator("legs")
    @classmethod
    def _unique_leg_ids(cls, legs: Optional[List[Leg]]) -> Optional[List[Leg]]:
        if legs is None:
            return legs
        ids = [leg.id for leg in legs]
        if len(set(ids)) != len(ids):
            raise ValueError("leg ids must be unique within a strategy")
        return legs

    @model_validator(mode="after")
    def _validate_tab_expiries(self) -> "StrategyUpdate":
        # Only validate when both legs and tab are being patched together;
        # otherwise the existing strategy.universe_tab governs and the
        # router-layer merge will re-validate. We don't have the existing
        # tab here, so be conservative: when both are present, check.
        if self.legs is not None and self.universe_tab is not None:
            _validate_tab_expiries(self.universe_tab, self.legs)
        if self.legs is not None and self.product is not None:
            _validate_product_vs_segments(self.product, self.legs)
        return self


# ----------------------------------------------------------------------------
# Output (read)
# ----------------------------------------------------------------------------


def _decimal_to_float(v):
    """Pydantic serializer helper for Numeric columns coming back as Decimal."""
    if v is None:
        return None
    return float(v)


class StrategyOut(BaseModel):
    """Detail / single-fetch response. Webhook URL is included; token is NOT."""

    model_config = ConfigDict(extra="forbid")

    id: int
    name: str
    strategy_kind: str = "batch"
    direction: str = "both"
    universe_tab: str
    underlying: str
    underlying_exchange: str
    strategy_type: str
    entry_time: Optional[str]
    exit_time: Optional[str]
    product: str
    pricetype: str
    legs: List[Leg]
    overall_sl_mtm: Optional[float]
    overall_target_mtm: Optional[float]
    lock_profit: Optional[LockProfitConfig]
    trail_sl_to_entry: bool
    scheduler: Optional[SchedulerConfig]

    live_enabled: bool
    webhook_url: str = Field(..., description="Public webhook URL for this strategy (token embedded)")
    webhook_ip_allowlist: Optional[List[WebhookIpAllowlistEntry]]
    webhook_locked: bool = False
    daily_loss_limit_inr: Optional[float]

    status: str
    current_run_id: Optional[int]

    created_at: str
    updated_at: str


class StrategyListItem(BaseModel):
    """Lighter shape for list view — drops legs/scheduler/etc.

    The list page in v1 still wants P&L columns (Section 3.1) but they require
    the engine which doesn't exist yet. Phase 1 returns realized/unrealized/total
    as 0 placeholders; Phase 6 wires them to live state.
    """

    model_config = ConfigDict(extra="forbid")

    id: int
    name: str
    strategy_kind: str = "batch"
    direction: str = "both"
    universe_tab: str
    underlying: str
    strategy_type: str
    status: str
    live_enabled: bool
    webhook_locked: bool = False

    pnl_realized: float = 0.0
    pnl_unrealized: float = 0.0
    pnl_total: float = 0.0

    created_at: str
    updated_at: str


class StrategyCreateResponse(BaseModel):
    """Special response for create + rotate: includes one-time-view webhook token."""

    model_config = ConfigDict(extra="forbid")

    strategy: StrategyOut
    webhook_token: str = Field(
        ...,
        description=(
            "PLAINTEXT webhook token. Shown ONCE — copy it now. "
            "It is hashed in the DB and cannot be retrieved later."
        ),
    )

File: frontend/src/types/strategy_module.ts
/**
 * Strategy module types — mirrors backend/schemas/strategy_module.py.
 * Names kept identical to the Pydantic schemas so a future codegen pass
 * can replace this file with generated bindings.
 */

export type UniverseTab =
  | "weekly_monthly"
  | "monthly_only"
  | "stocks_fno"
  | "mcx";

export type StrategyType = "intraday" | "positional";
export type StrategyStatus = "stopped" | "running" | "paused" | "errored";
export type StrategyMode = "live" | "sandbox";
export type Product = "NRML" | "MIS" | "CNC";
export type PriceType = "MARKET" | "LIMIT";
export type Position = "B" | "S";
export type OptionType = "CE" | "PE";
export type Segment = "options" | "futures" | "cash";
/**
 * Canonical expiry ranks (preferred):
 *   - current_week  — nearest expiry (weekly or monthly, whichever is rank-1)
 *   - next_week     — second-nearest expiry
 *   - current_month — nearest "last-of-calendar-month" expiry
 *   - next_month    — month after current_month
 *
 * Legacy aliases (kept for backwards compatibility with rows created
 * before the explicit-ranks rollout): weekly ≡ current_week,
 * monthly ≡ current_month, current ≡ current_month, next ≡ next_month.
 * Backend resolves both old and new names; new strategies use the
 * canonical four.
 */
export type ExpiryRank =
  | "current_week"
  | "next_week"
  | "current_month"
  | "next_month"
  | "weekly"
  | "monthly"
  | "current"
  | "next";
export type StrikeMode = "atm" | "strike";
export type Weekday = "MON" | "TUE" | "WED" | "THU" | "FRI" | "SAT" | "SUN";

/**
 * Signal-mode discriminator. 'batch' = multi-leg options spread (existing).
 * 'signal' = TradingView-style per-leg long/short entry/exit signals.
 * Default 'batch' for backwards compatibility with rows created before the
 * signal-mode schema migration (slice 2).
 */
export type StrategyKind = "batch" | "signal";

/**
 * Direction filter for signal-mode strategies. Ignored for batch-mode.
 * The engine and webhook handler both gate incoming signals against this
 * (slices 3 and 5). 'both' = no filtering.
 */
export type StrategyDirection = "long_only" | "short_only" | "both";

/**
 * Per-leg side declaration for signal-mode legs. Determines which signal
 * actions a leg accepts:
 *   - 'long'  -> long_entry + long_exit
 *   - 'short' -> short_entry + short_exit
 *   - 'both'  -> all four actions
 * Unrelated to batch-mode's per-leg `position` (B/S).
 */
export type LegSide = "long" | "short" | "both";

/** The four TradingView-style signal actions for signal-mode webhooks. */
export type SignalAction =
  | "long_entry"
  | "long_exit"
  | "short_entry"
  | "short_exit";

export interface TrailConfig {
  x: number;
  y: number;
}

export interface Leg {
  id: number;
  segment: Segment;
  // Optional in signal mode (cash legs have no expiry; futures use
  // 'current' | 'next'). Required for batch-mode F&O legs - the backend
  // Pydantic validator enforces this per-kind.
  expiry?: ExpiryRank | null;
  // Batch-mode lot count multiplied by symtoken.lotsize at runtime.
  // Default 1 for backwards compat with rows that don't set it.
  lots: number;
  position: Position;
  option_type?: OptionType | null;
  strike_mode?: StrikeMode | null;
  atm_offset?: string | null;
  strike_value?: number | null;

  // --- Signal-mode fields (null/undefined for batch-mode legs) ---
  symbol?: string | null;
  exchange?: string | null;
  side?: LegSide | null;
  /** Absolute order quantity for signal-mode legs (shares for cash,
   *  lot-multiple for futures - the wizard handles the multiplication
   *  at form-submit time). Replaces `lots` as the quantity authority
   *  for signal-mode strategies. */
  qty?: number | null;

  target_pts?: number | null;
  sl_pts?: number | null;
  trail: TrailConfig;
  momentum?: Record<string, unknown> | null;
}

export interface LockProfitConfig {
  mode: "lock" | "lock_and_trail";
  if_profit_reaches: number;
  lock_profit: number;
  trail_step?: number | null;
}

export interface SchedulerConfig {
  enabled: boolean;
  days: Weekday[];
  start_time: string; // HH:MM
  auto_stop_time?: string | null;
  default_mode: StrategyMode;
}

export interface WebhookIpAllowlistEntry {
  cidr: string;
  label?: string | null;
}

export interface StrategyCreate {
  name: string;
  /** 'batch' (default) keeps the existing multi-leg options spread flow.
   *  'signal' opts in to the TradingView per-leg signal flow. The
   *  backend's strategy_kind column defaults to 'batch' so omitting
   *  this field preserves current wizard behaviour. */
  strategy_kind?: StrategyKind;
  /** Signal-mode direction filter. Ignored for batch-mode strategies
   *  but always sent so the wire format is consistent. */
  direction?: StrategyDirection;
  universe_tab: UniverseTab;
  underlying: string;
  underlying_exchange: string;
  strategy_type: StrategyType;
  entry_time?: string | null;
  exit_time?: string | null;
  product: Product;
  pricetype: PriceType;
  legs: Leg[];
  overall_sl_mtm?: number | null;
  overall_target_mtm?: number | null;
  lock_profit?: LockProfitConfig | null;
  trail_sl_to_entry: boolean;
  scheduler?: SchedulerConfig | null;
  webhook_ip_allowlist?: WebhookIpAllowlistEntry[] | null;
  daily_loss_limit_inr?: number | null;
}

/** Partial update for PATCH /web/strategy/{id}.
 *  Note: strategy_kind is intentionally omitted - the backend Pydantic
 *  schema also forbids it (immutable post-create per slice 2 design).
 *  Direction can be patched freely. */
export type StrategyUpdate = Partial<Omit<StrategyCreate, "strategy_kind">>;

export interface Strategy {
  id: number;
  name: string;
  /** Defaults to 'batch' when the server didn't ship the field (old DB row). */
  strategy_kind: StrategyKind;
  direction: StrategyDirection;
  universe_tab: UniverseTab;
  underlying: string;
  underlying_exchange: string;
  strategy_type: StrategyType;
  entry_time?: string | null;
  exit_time?: string | null;
  product: Product;
  pricetype: PriceType;
  legs: Leg[];
  overall_sl_mtm?: number | null;
  overall_target_mtm?: number | null;
  lock_profit?: LockProfitConfig | null;
  trail_sl_to_entry: boolean;
  scheduler?: SchedulerConfig | null;
  live_enabled: boolean;
  webhook_url: string;
  webhook_ip_allowlist?: WebhookIpAllowlistEntry[] | null;
  /** Kill-switch flag. When true, the webhook handler refuses every
   *  incoming signal until /unlock_webhook is called. */
  webhook_locked: boolean;
  daily_loss_limit_inr?: number | null;
  status: StrategyStatus;
  current_run_id?: number | null;
  created_at: string;
  updated_at: string;
}

export interface StrategyListItem {
  id: number;
  name: string;
  strategy_kind: StrategyKind;
  direction: StrategyDirection;
  universe_tab: UniverseTab;
  underlying: string;
  strategy_type: StrategyType;
  status: StrategyStatus;
  live_enabled: boolean;
  webhook_locked: boolean;
  pnl_realized: number;
  pnl_unrealized: number;
  pnl_total: number;
  created_at: string;
  updated_at: string;
}

export interface StrategyCreateResponse {
  strategy: Strategy;
  webhook_token: string;
}

export const UNIVERSE_TAB_LABELS: Record<UniverseTab, string> = {
  weekly_monthly: "Weekly & Monthly Expiries",
  monthly_only: "Monthly Only Expiry",
  stocks_fno: "Stocks – Cash / F&O",
  mcx: "Commodities (MCX)",
};

export const UNIVERSE_TAB_HINT: Record<UniverseTab, string> = {
  weekly_monthly: "NIFTY, SENSEX",
  monthly_only: "MIDCPNIFTY, BANKNIFTY, FINNIFTY, BANKEX",
  stocks_fno: "All NIFTY 500 stocks",
  mcx: "CRUDEOIL, NATURALGAS, GOLD, SILVER, …",
};

/**
 * What expiry ranks the wizard offers per tab. Backend validator is the
 * authority and rejects mismatches (e.g. weekly on MCX).
 *
 * Index tabs (NIFTY / SENSEX) get the full four since they have both
 * weekly + monthly contracts. Stock F&O is monthly-only; so is MCX.
 */
export const TAB_EXPIRIES: Record<UniverseTab, ExpiryRank[]> = {
  weekly_monthly: ["current_week", "next_week", "current_month", "next_month"],
  monthly_only: ["current_month", "next_month"],
  stocks_fno: ["current_month", "next_month"],
  mcx: ["current_month", "next_month"],
};

/** Human-readable labels for the expiry-rank dropdown. */
export const EXPIRY_RANK_LABELS: Record<ExpiryRank, string> = {
  current_week: "Current Week",
  next_week: "Next Week",
  current_month: "Current Month",
  next_month: "Next Month",
  // Legacy aliases — kept so existing strategies render sensibly in the
  // edit form while they migrate to the canonical names on next save.
  weekly: "Current Week (legacy)",
  monthly: "Current Month (legacy)",
  current: "Current Month (legacy)",
  next: "Next Month (legacy)",
};

/** What segments are allowed per tab. */
export const TAB_SEGMENTS: Record<UniverseTab, Segment[]> = {
  weekly_monthly: ["futures", "options"],
  monthly_only: ["futures", "options"],
  stocks_fno: ["cash", "futures", "options"],
  mcx: ["futures", "options"],
};

/**
 * Default underlyings per tab — used as the dropdown source until Phase 3
 * wires the dynamic /api/v1/strategy/underlyings endpoint. The MCX list is
 * Phase 3-dynamic; for Phase 2 we just take a small known-good seed.
 */
export const TAB_DEFAULT_UNDERLYINGS: Record<
  UniverseTab,
  Array<{ symbol: string; name: string; exchange: string }>
> = {
  weekly_monthly: [
    { symbol: "NIFTY", name: "Nifty 50", exchange: "NSE_INDEX" },
    { symbol: "SENSEX", name: "BSE SENSEX", exchange: "BSE_INDEX" },
  ],
  monthly_only: [
    { symbol: "BANKNIFTY", name: "Nifty Bank", exchange: "NSE_INDEX" },
    { symbol: "FINNIFTY", name: "Nifty Fin Service", exchange: "NSE_INDEX" },
    { symbol: "MIDCPNIFTY", name: "Nifty Midcap Select", exchange: "NSE_INDEX" },
    { symbol: "BANKEX", name: "BSE Bankex", exchange: "BSE_INDEX" },
  ],
  stocks_fno: [
    { symbol: "RELIANCE", name: "Reliance Industries", exchange: "NSE" },
    { symbol: "TCS", name: "Tata Consultancy Services", exchange: "NSE" },
    { symbol: "HDFCBANK", name: "HDFC Bank", exchange: "NSE" },
    { symbol: "INFY", name: "Infosys", exchange: "NSE" },
  ],
  mcx: [
    { symbol: "CRUDEOIL", name: "Crude Oil", exchange: "MCX" },
    { symbol: "NATURALGAS", name: "Natural Gas", exchange: "MCX" },
    { symbol: "GOLD", name: "Gold", exchange: "MCX" },
    { symbol: "SILVER", name: "Silver", exchange: "MCX" },
  ],
};

/** ATM offset choices shown in the strike-criteria dropdown. */
export const ATM_OFFSETS: string[] = [
  "ATM",
  "ITM1",
  "ITM2",
  "ITM3",
  "ITM4",
  "ITM5",
  "OTM1",
  "OTM2",
  "OTM3",
  "OTM4",
  "OTM5",
];

// ---------------------------------------------------------------------------
// Signal-mode UX constants (slice 7)
// ---------------------------------------------------------------------------

/** Human-readable labels for the strategy-kind picker at the top of the
 *  wizard (design section 6.1). */
export const STRATEGY_KIND_LABELS: Record<StrategyKind, string> = {
  batch: "Multi-leg (batch)",
  signal: "Signal-driven (TradingView)",
};

/** One-line hint shown under the kind picker so the user knows which
 *  is which without diving into docs. */
export const STRATEGY_KIND_HINT: Record<StrategyKind, string> = {
  batch:
    "All legs entered together on start; exited together on stop. Best for option spreads.",
  signal:
    "Each leg reacts to long_entry / long_exit / short_entry / short_exit signals.",
};

/** Labels for the direction radio. */
export const STRATEGY_DIRECTION_LABELS: Record<StrategyDirection, string> = {
  long_only: "Long only",
  short_only: "Short only",
  both: "Both",
};

/** Labels for the per-leg side selector in the signal-mode wizard. */
export const LEG_SIDE_LABELS: Record<LegSide, string> = {
  long: "Long",
  short: "Short",
  both: "Both",
};

/** Universe tabs allowed when the strategy is signal-mode. All four tabs
 *  are supported: index-options (weekly_monthly, monthly_only), stock
 *  cash/futures/options (stocks_fno), and MCX futures/options. The
 *  engine resolves each leg's contract from (symbol, expiry rank,
 *  option fields) at signal time so the leg config stays roll-stable. */
export const SIGNAL_MODE_ALLOWED_TABS: UniverseTab[] = [
  "weekly_monthly",
  "monthly_only",
  "stocks_fno",
  "mcx",
];

/** Default product per (strategy_type, segment) for signal-mode legs.
 *  Wizard auto-picks unless the user overrides. */
export function defaultProductForSignal(
  strategyType: StrategyType,
  segment: Segment,
): Product {
  if (segment === "cash") {
    return strategyType === "intraday" ? "MIS" : "CNC";
  }
  // futures/options: MIS for intraday, NRML for positional
  return strategyType === "intraday" ? "MIS" : "NRML";
}

// ---------------------------------------------------------------------------
// Tab / segment -> Product rules (slice 11)
// ---------------------------------------------------------------------------
//
// Exchange-side product rules:
//   - Cash equity (NSE/BSE)        -> MIS or CNC      (default MIS)
//   - Derivatives (NFO/BFO/MCX)    -> NRML or MIS     (default NRML)
//
// The strategy-level product applies to every leg, so for a strategy
// that mixes cash + derivatives (only possible on stocks_fno) we
// restrict to MIS (the only product valid in both contexts).
// ---------------------------------------------------------------------------

/** Products valid for the given mix of segments across legs. */
export function allowedProductsForLegs(legs: Leg[]): Product[] {
  const segs = new Set(legs.map((l) => l.segment));
  const hasCash = segs.has("cash");
  const hasDeriv = segs.has("futures") || segs.has("options");
  if (hasCash && hasDeriv) return ["MIS"];
  if (hasCash) return ["MIS", "CNC"];
  return ["NRML", "MIS"];
}

/** Default product for a strategy given its leg composition. Cash-only
 *  defaults to MIS; derivatives default to NRML. */
export function defaultProductForLegs(legs: Leg[]): Product {
  const allowed = allowedProductsForLegs(legs);
  return allowed[0];
}

// ---------------------------------------------------------------------------
// Tab -> Intraday default windows (slice 11)
// ---------------------------------------------------------------------------
//
// NSE/BSE cash + F&O trade 09:15-15:30 IST. Operator-friendly intraday
// window is 09:35-15:15 (skip opening volatility, exit before MIS auto-
// squareoff at 15:20 from most brokers).
//
// MCX commodities trade 09:00-23:30 (winter) / 23:55 (summer). Default
// intraday window is 09:00-23:25 so positions auto-square comfortably
// before the session close.
// ---------------------------------------------------------------------------

export const TAB_INTRADAY_DEFAULTS: Record<
  UniverseTab,
  { entry: string; exit: string }
> = {
  weekly_monthly: { entry: "09:35", exit: "15:15" },
  monthly_only: { entry: "09:35", exit: "15:15" },
  stocks_fno: { entry: "09:35", exit: "15:15" },
  mcx: { entry: "09:00", exit: "23:25" },
};
